src/dataService/listPaginationService.ts

```diff
@@ -0,0 +1,657 @@
+   * some shared temp table logic.
+   */
+  private async paginatedResult(
+    query: Knex.QueryBuilder,
+    trx: Knex.Transaction,
+    pagination: PaginationInput,
+    sort: SavedItemsSort
+  ) {
+    const queryBuilder = query.select(
+      'list.item_id',
+      'list.resolved_id',
+      'list.given_url',
+      'list.title',
+      'list.favorite',
+      'list.status',
+      trx.raw('UNIX_TIMESTAMP(list.time_added) as time_added'),
+      trx.raw('UNIX_TIMESTAMP(list.time_updated) AS time_updated'),
+      trx.raw('UNIX_TIMESTAMP(list.time_read) AS time_read'),
+      trx.raw('UNIX_TIMESTAMP(list.time_favorited) AS time_favorited')
+    );
+    // needs to be same order as above
+    const insertStatement = `INSERT INTO \`${this.tempTable}\` (item_id, resolved_id, given_url, given_title, favorite, status, time_added, time_updated, time_read, time_favorited) `;
+    const cursor = pagination.after ?? pagination.before ?? null;
+    if (cursor) {
+      return this.pageAfterorBefore(
+        trx,
+        queryBuilder,
+        insertStatement,
+        cursor,
+        pagination,
+        sort
+      );
+    } else {
+      return this.pageFirstLast(
+        trx,
+        queryBuilder,
+        insertStatement,
+        sort,
+        pagination
+      );
+    }
+  }
+
+  /**
+   * Handle first/last pagination.
+   */
+  private async pageFirstLast(
+    trx: Knex.Transaction,
+    query: Knex.QueryBuilder,
+    insertStatement: string,
+    sort: SavedItemsSort,
+    pagination: PaginationInput
+  ) {
+    const pageSize = pagination.first ?? pagination.last;
+    const sortOrder = new Sort(sort);
+    let order;
+    if (pagination.first) {
+      order = sortOrder.order;
+    } else {
+      order = sortOrder.opposite;
+    }
+    const sortColumn = this.sortMap[sortOrder.column];
+    const queryString = query
+      .clone()
+      .orderBy([
+        { column: `list.${sortColumn}`, order: order },
+        { column: 'list.item_id' },
+      ])
+      .limit(pageSize + 1)
+      .toString();
+    await trx.raw(`${insertStatement} ${queryString}`);
+    const returnQuery = trx(this.tempTable).select();
+    if (pagination.last) {
+      // Need to reorder for last
+      returnQuery.orderBy([
+        { column: `${sortColumn}`, order: sortOrder.order },
+        { column: 'item_id' },
+      ]);
+    }
+    return await returnQuery;
+  }
+
+  /**
+   * Handle before/after pagination.
+   * If the provided cursor does not exist, throws UserInputError.
+   */
+  private async pageAfterorBefore(
+    trx: Knex.Transaction,
+    baseQuery: Knex.QueryBuilder,
+    insertStatement: string,
+    cursor: string,
+    pagination: PaginationInput,
+    sort: SavedItemsSort
+  ) {
+    const pageSize = pagination.first ?? pagination.last;
+    // Since we don't have a unique sequential column for cursor-based pagination
+    // We have to get the old cursor element + any colliding keys
```

Comment:
quick question on `collission`. could you give me an example of collission with cursor

how i am understanding is: the cursor is `itemId+epoc` but for the single user. say the sort column is `updatedAt`, the cursor is `itemId+updatedAt`, and this might not collide with other item in the list as they have a different itemId.

Reply:
sure! let's do `favoritedAt` sorted `asc`.

let's say the cursor decodes to `5_*_null`.

if there weren't any collisions (say itemId 5 was the only null value in the db), then we'd have a sort something like

```
5 | null
2 | <timestamp>
1 | <timestamp>+n // some later timestamp
6 | <timestamp>+m
...
```

This is good. The first row contains the cursor value, which is what we expect.

but if there's collisions, then our sort falls back on itemId, ascending:

```
1 | null
2 | null
5 | null
6 | null
...
```

then the first row is *not* our cursor value. And we can't just include a filter by itemId, because itemId sequence has no sequential relationship to any of the epoch timestamps. So we have to get all the collisions. Then in this case we'd strip off the first two rows, so now the cursor is the first row. Then we'd decide if we need to pull in more data or not (in the case where there are many collisions, we might have the entire page or more already).

Does that help?